### PR TITLE
Lf read From merlokk

### PR DESCRIPTION
**Pull request by xianglin1998**
**Strengthen low-frequency related functions**

### CODE REVIEW

File: software/script/chameleon_enum.py [hunk 1]
```diff
@@ -72,6 +72,7 @@ class Command(enum.IntEnum):
 
     EM410X_SCAN = 3000
     EM410X_WRITE_TO_T55XX = 3001
+    DATA_CMD_LF_READ = 3010
```
Comment (taichunmin) [inline software/script/chameleon_enum.py:75]: maybe rename to `LF_READ`?

File: software/script/chameleon_enum.py [hunk 2]
```diff
@@ -116,6 +117,8 @@ class Status(enum.IntEnum):
     LF_TAG_OK = 0x40
     # Unable to search for a valid EM410X label
     EM410X_TAG_NO_FOUND = 0x41
+    # Unable to search for a valid LF tag
+    LF_TAG_NO_FOUND = 0x42
```
Comment (taichunmin) [inline software/script/chameleon_enum.py:121]: maybe rename `EM410X_TAG_NO_FOUND` to `LF_TAG_NO_FOUND`?

Comment (xianglin1998) [reply software/script/chameleon_enum.py:121]: > maybe rename `EM410X_TAG_NO_FOUND` to `LF_TAG_NO_FOUND` and reuse `0x41`?

At present, this PR is in the testing phase, and all code formats and naming conventions will be corrected before merging.

**Comment by augustozanellato** [comment]
A notice should probably be added stating that this code is at least partly based on the Flipper lfrfid library

**Comment by shayya** [comment]
can we start working on HID26 protocol emulation. It's by far the most common LF protocol. I'm not as familiar with the chameleon ultra codebase. If you tell me what are the files that are need I'd be happy to contribute

**Comment by xianglin1998** [comment]
> A notice should probably be added stating that this code is at least partly based on the Flipper lfrfid library

I will confirm some copyright related issues with the author, which is indeed very important.

**Comment by xianglin1998** [comment]
> can we start working on HID26 protocol emulation. It's by far the most common LF protocol. I'm not as familiar with the chameleon ultra codebase. If you tell me what are the files that are need I'd be happy to contribute

Within the "firmware/application/src/rfid", you can see the '[nfctag](https://github.com/RfidResearchGroup/ChameleonUltra/tree/main/firmware/application/src/rfid/nfctag)' and '[reader](https://github.com/RfidResearchGroup/ChameleonUltra/tree/main/firmware/application/src/rfid/reader)' dir, if you are want to implement a tag,  you can view the nfctag dir's file.

**Comment by merlokk** [comment]
> > A notice should probably be added stating that this code is at least partly based on the Flipper lfrfid library
> 
> I will confirm some copyright related issues with the author, which is indeed very important.

yes, it GPL3. license will be added later.

**Comment by yarreg** [comment]
What is the status of this PR? We are waiting for a new code base to support more LF protocols. Thanks.

**Comment by merlokk** [comment]
im working on it. i think in several weeks I will publish it

**Comment by yarreg** [comment]
> im working on it. i think in several weeks I will publish it
Thanks for your work!

**Comment by Xample** [comment]
Hello, here is my feedback:
I recently purchased a Chameleon Ultra which one I immediately updated with the latest FW but reading the LF never worked. I first thought the Chameleon to be faulty and therefore purchased another one supposed to be genuine. On delivery, it was loaded with the FW 1.0, I tried loading a LF card and the new Chameleon was able to detect it. I then tried to upgrade it to the v2.0.0 but then my Chameleon is not reading LF anymore. I saw many users experienced the same issue (which is the reason why I'm here). I therefore tried your firmware (based on the artefact) on at least one device but, it is still not reading any card. I therefore rebased your branch on master and rebuilt the firmware. Something like

```bash
git clone https://github.com/merlokk/ChameleonUltra.git
cd ChameleonUltra
git remote add upstream https://github.com/RfidResearchGroup/ChameleonUltra.git
git fetch upstream
git checkout lf_read
git rebase upstream/main
docker run --rm --platform=linux/amd64 -v "$(pwd)":/workdir -w /workdir ghcr.io/rfidresearchgroup/chameleonultra-fw-builder:pr-233 bash -c "./firmware/build.sh"
```
For the curious, the builds will be in `ChameleonUltra/firmware/objects/bootloader`
I then updated my 2 Chameleons (using MTools BLE using my iPhone), but I'm still unable to read any `lf` card.

I therefore think about downgrading the FW to 1.0 to start testing my Chameleons with a firmware from scratch. But having no XXXLink [(based on the only artefact available for 1.0)](https://github.com/RfidResearchGroup/ChameleonUltra/releases/tag/v1.0) I just don't know how to load or pack this [ultra.hex](https://github.com/RfidResearchGroup/ChameleonUltra/releases/download/v1.0/ultra.hex) to load it into my chameleon. I will open a ticket to ask for help.

To summarize: My issue has not been fixed (yet?) by this branch's code. I will continue digging on my side to see if my device is "corrupted" and if so, why.

**Comment by whywilson** [comment]
> Hello, here is my feedback: I recently purchased a Chameleon Ultra which one I immediately updated with the latest FW but reading the LF never worked. I first thought the Chameleon to be faulty and therefore purchased another one supposed to be genuine. On delivery, it was loaded with the FW 1.0, I tried loading a LF card and the new Chameleon was able to detect it. I then tried to upgrade it to the v2.0.0 but then my Chameleon is not reading LF anymore. I saw many users experienced the same issue (which is the reason why I'm here). I therefore tried your firmware (based on the artefact) on at least one device but, it is still not reading any card. I therefore rebased your branch on master and rebuilt the firmware. Something like
> 
> ```shell
> git clone https://github.com/merlokk/ChameleonUltra.git
> cd ChameleonUltra
> git remote add upstream https://github.com/RfidResearchGroup/ChameleonUltra.git
> git fetch upstream
> git checkout lf_read
> git rebase upstream/main
> docker run --rm --platform=linux/amd64 -v "$(pwd)":/workdir -w /workdir ghcr.io/rfidresearchgroup/chameleonultra-fw-builder:pr-233 bash -c "./firmware/build.sh"
> ```
> 
> For the curious, the builds will be in `ChameleonUltra/firmware/objects/bootloader` I then updated my 2 Chameleons (using MTools BLE using my iPhone), but I'm still unable to read any `lf` card.
> 
> I therefore think about downgrading the FW to 1.0 to start testing my Chameleons with a firmware from scratch. But having no XXXLink [(based on the only artefact available for 1.0)](https://github.com/RfidResearchGroup/ChameleonUltra/releases/tag/v1.0) I just don't know how to load or pack this [ultra.hex](https://github.com/RfidResearchGroup/ChameleonUltra/releases/download/v1.0/ultra.hex) to load it into my chameleon. [I will open a ticket to ask for help.](https://github.com/RfidResearchGroup/ChameleonUltra/issues/236)
> 
> To summarize: My issue has not been fixed (yet?) by this branch's code. I will continue digging on my side to see if my device is "corrupted" and if so, why.

If you have the issues below, it can be the antenna issues.
FW 1.0 return all 00 when run `lf read` no matter if there is a EM4100 tag there or not. FW 2.0 doesn't return anyting.

**Comment by GameTec-live** [comment]
Whats the state on this PR? Still stuff going on or?

**Comment by TeCHiScy** [comment]
> Whats the state on this PR? Still stuff going on or?

My code for HIDProx starts from this PR, and in the journey, things that over-designed (like bit array/level duration) was stripped. Thus, imo, this PR can be considered supersede by the #267.

**Comment by merlokk** [comment]
For now, it is over-engineered. I considered adapting code from one of the open-source projects, but the reader is too different.
needs to make some code from scratch(
